File: app/src/main/java/net/ballmerlabs/uscatterbrain/network/InputStreamObserver.java
package net.ballmerlabs.uscatterbrain.network;

import java.io.IOException;

import io.reactivex.Observer;
import io.reactivex.disposables.Disposable;

public class InputStreamObserver extends InputStreamCallback implements Observer<byte[]> {

    public InputStreamObserver(int capacity) {
        super(capacity);
    }

    public InputStreamObserver() {
        super();
    }

    @Override
    public void onSubscribe(Disposable d) {
        disposable = d;
    }

    @Override
    public void onNext(byte[] bytes) {
        if (!closed) {
            acceptBytes(bytes);
        }
    }

    @Override
    public void onError(Throwable e) {
        throwable = e;
        try {
            close();
        } catch (IOException ignored) {

        }
    }

    @Override
    public void onComplete() {

    }
}

File: app/src/main/java/net/ballmerlabs/uscatterbrain/network/InputStreamObserver.kt
package net.ballmerlabs.uscatterbrain.network

import io.reactivex.Observer
import io.reactivex.disposables.Disposable
import java.io.IOException

/**
 * serves as a bridge between an observable and an inputstream
 * using a static sized buffer to hold unread data
 */
class InputStreamObserver(capacity: Int) : InputStreamCallback(capacity), Observer<ByteArray> {

    override fun onSubscribe(d: Disposable) {

    }

    override fun onNext(bytes: ByteArray) {
        if (!closed) {
            acceptBytes(bytes)
        }
    }

    override fun onError(e: Throwable) {
        throwable = e
        try {
            close()
        } catch (ignored: IOException) {
        }
    }

    override fun onComplete() {
    }
}